worker/uniter/runner/context/context.go

```diff
@@ -486,13 +487,39 @@ func (ctx *HookContext) ConfigSettings() (charm.Settings, error) {
 	return result, nil
 }
 
-func (ctx *HookContext) GoalState() (string, error) {
-	var err error
-	ctx.goalState, err = ctx.state.GoalState()
+func (ctx *HookContext) GoalState() (*application.GoalState, error) {
+	paramsGoalState, err := ctx.state.GoalState()
 	if err != nil {
-		return "", err
+		return nil, err
+	}
+	ctx.goalState = transformGoalState(&paramsGoalState)
+
+	return &ctx.goalState, nil
+}
+
+// transformGoalState move information from params GoalState struct to
+// application GoalState struct
+func transformGoalState(api *uniter.GoalState) application.GoalState {
```

Comment:
This is good but typically we would look to do this transformation in the api layer. In theory, params structs should not escape from the api layer and all api calls would return structs from core. We don't do that consistently sadly. But we can do it correctly here. Just move this method the the api layer and change the return type of api/uniter.GoalState()